### PR TITLE
Roe 1432 correct combination bo mo

### PR DESCRIPTION
**Pull request by mattch1**
https://companieshouse.atlassian.net/browse/ROE-1304

Refactored the due diligence and officers code.
For the owners and officers check whether any of the lists are not null and not empty and if so return true. Then the conditions are assed for each statement type based on whether they should or should not be present and a specific message is returned based on the statement and the error found.

The tests cover all the AC from 7-13 and an additional test to ensure that when none are identified that at least one managing officer is present.

### CODE REVIEW

File: src/main/java/uk/gov/companieshouse/overseasentitiesapi/validation/utils/OwnersAndOfficersDataBlockValidator.java [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+                if (!hasManagingOfficers(overseasEntitySubmissionDto)) {
+                    logValidationErrorMessage(errors, loggingContext, String.format("%s for statement that none can be identified", MISSING_MANAGING_OFFICER));
+                    return false;
+                }
+                if (hasBeneficialOwners(overseasEntitySubmissionDto)) {
+                    logValidationErrorMessage(errors, loggingContext, String.format("%s for statement that none can be identified", INCORRECTLY_ADDED_BENEFICIAL_OWNER));
+                    return false;
+                }
+                break;
+        }
```
Comment (markpit) [inline src/main/java/uk/gov/companieshouse/overseasentitiesapi/validation/utils/OwnersAndOfficersDataBlockValidator.java:111]: if the user supplies a non standard value for the statement eg "NONE", would it pass the validation? I think it would be good to have a 'default' in the switch so that if the value is not matched against our expected statement values then we should return validation message - 'Invalid statement supplied'?

Possibly add a similar check to the beneficialOwnersStatementValidator.validate() to block any random statement values (if it doesn't already)

File: src/main/java/uk/gov/companieshouse/overseasentitiesapi/validation/OverseasEntitySubmissionDtoValidator.java [hunk 2]
```diff
@@ -2,58 +2,31 @@
 
 import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.stereotype.Component;
-import uk.gov.companieshouse.overseasentitiesapi.model.dto.BeneficialOwnerCorporateDto;
-import uk.gov.companieshouse.overseasentitiesapi.model.dto.BeneficialOwnerGovernmentOrPublicAuthorityDto;
-import uk.gov.companieshouse.overseasentitiesapi.model.dto.BeneficialOwnerIndividualDto;
-import uk.gov.companieshouse.overseasentitiesapi.model.dto.ManagingOfficerCorporateDto;
-import uk.gov.companieshouse.overseasentitiesapi.model.dto.ManagingOfficerIndividualDto;
 import uk.gov.companieshouse.overseasentitiesapi.model.dto.OverseasEntitySubmissionDto;
 import uk.gov.companieshouse.overseasentitiesapi.validation.utils.DueDiligenceDataBlockValidator;
```
Comment (markpit) [inline src/main/java/uk/gov/companieshouse/overseasentitiesapi/validation/OverseasEntitySubmissionDtoValidator.java:6]: i'd move the DueDiligenceDataBlockValidator and OwnersAndOfficersDataBlockValidator out of the utils package and into the validation package as they are business logic related. Utils is more for things like string checks etc that have a lot of re-use.

File: src/test/java/uk/gov/companieshouse/overseasentitiesapi/validation/OverseasEntitySubmissionDtoValidatorTest.java [hunk 3]
```diff
@@ -51,25 +51,11 @@ class OverseasEntitySubmissionDtoValidatorTest {
...
-    private BeneficialOwnerGovernmentOrPublicAuthorityValidator beneficialOwnerGovernmentOrPublicAuthorityValidator;
-    @Mock
     private DueDiligenceDataBlockValidator dueDiligenceDataBlockValidator;
     @Mock
-    private ManagingOfficerIndividualValidator managingOfficerIndividualValidator;
-    @Mock
-    private ManagingOfficerCorporateValidator managingOfficerCorporateValidator;
+    private OwnersAndOfficersDataBlockValidator ownersAndOfficersDataBlockValidator;
 
     private EntityDto entityDto = EntityMock.getEntityDto();
```
Comment (markpit) [inline src/test/java/uk/gov/companieshouse/overseasentitiesapi/validation/OverseasEntitySubmissionDtoValidatorTest.java:60]: intellij has highlighted most of the private dto class level variables and recommends making them 'final'.

File: src/test/java/uk/gov/companieshouse/overseasentitiesapi/validation/utils/DueDiligenceDataBlockValidatorTest.java [hunk 4]
```diff
@@ -1,110 +1,122 @@
...
+   @InjectMocks
+   private DueDiligenceDataBlockValidator dueDiligenceDataBlockValidator;
+   @Mock
+   private OverseasEntityDueDiligenceValidator overseasEntityDueDiligenceValidator;
+   @Mock
+   private DueDiligenceValidator dueDiligenceValidator;
 
-   private DueDiligenceDataBlockValidator dueDiligenceDataBlockValidator = new DueDiligenceDataBlockValidator();
    private OverseasEntityDueDiligenceDto overseasEntityDueDiligenceDto = OverseasEntityDueDiligenceMock.getOverseasEntityDueDiligenceDto();
    private DueDiligenceDto dueDiligenceDto = DueDiligenceMock.getDueDiligenceDto();
```
Comment (markpit) [inline src/test/java/uk/gov/companieshouse/overseasentitiesapi/validation/utils/DueDiligenceDataBlockValidatorTest.java:33]: OverseasEntityDueDiligenceDto and dueDiligenceDto can be 'final'

File: src/test/java/uk/gov/companieshouse/overseasentitiesapi/validation/utils/DueDiligenceDataBlockValidatorTest.java [hunk 5]
```diff
@@ -1,110 +1,122 @@
...
```
Comment (markpit) [inline src/test/java/uk/gov/companieshouse/overseasentitiesapi/validation/utils/DueDiligenceDataBlockValidatorTest.java:34]: I'd also add 2 tests to verify that dueDiligenceDataBlockValidator or overseasEntityDueDiligenceValidator gets called when either one of those blocks is present.

File: src/test/java/uk/gov/companieshouse/overseasentitiesapi/validation/utils/OwnersAndOfficersDataBlockValidatorTest.java [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+    private BeneficialOwnerGovernmentOrPublicAuthorityValidator beneficialOwnerGovernmentOrPublicAuthorityValidator;
+    @Mock
+    private ManagingOfficerIndividualValidator managingOfficerIndividualValidator;
+    @Mock
+    private ManagingOfficerCorporateValidator managingOfficerCorporateValidator;
+
+    private OverseasEntitySubmissionDto overseasEntitySubmissionDto;
+    private EntityDto entityDto = EntityMock.getEntityDto();
+    private PresenterDto presenterDto = PresenterMock.getPresenterDto();
+    private DueDiligenceDto dueDiligenceDto = DueDiligenceMock.getDueDiligenceDto();
```
Comment (markpit) [inline src/test/java/uk/gov/companieshouse/overseasentitiesapi/validation/utils/OwnersAndOfficersDataBlockValidatorTest.java:65]: these dto vars can be 'final'

File: src/test/java/uk/gov/companieshouse/overseasentitiesapi/validation/utils/OwnersAndOfficersDataBlockValidatorTest.java [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+
```
Comment (markpit) [inline src/test/java/uk/gov/companieshouse/overseasentitiesapi/validation/utils/OwnersAndOfficersDataBlockValidatorTest.java:66]: I'd also add a test or tests to verify the correct BO or MO validator gets called if the BO or Mo is present

File: src/main/java/uk/gov/companieshouse/overseasentitiesapi/validation/utils/OwnersAndOfficersDataBlockValidator.java (hunk 1 above)
Comment (mattch1) [reply src/main/java/uk/gov/companieshouse/overseasentitiesapi/validation/utils/OwnersAndOfficersDataBlockValidator.java:111]: As this is an enum I'm not sure this is going to help us.

File: src/main/java/uk/gov/companieshouse/overseasentitiesapi/validation/OverseasEntitySubmissionDtoValidator.java (hunk 2 above)
Comment (mattch1) [reply src/main/java/uk/gov/companieshouse/overseasentitiesapi/validation/OverseasEntitySubmissionDtoValidator.java]: Moved these

File: src/test/java/uk/gov/companieshouse/overseasentitiesapi/validation/OverseasEntitySubmissionDtoValidatorTest.java (hunk 3 above)
Comment (mattch1) [reply src/test/java/uk/gov/companieshouse/overseasentitiesapi/validation/OverseasEntitySubmissionDtoValidatorTest.java]: Added final same for below

File: src/main/java/uk/gov/companieshouse/overseasentitiesapi/validation/utils/OwnersAndOfficersDataBlockValidator.java (hunk 1 above)
Comment (markpit) [reply src/main/java/uk/gov/companieshouse/overseasentitiesapi/validation/utils/OwnersAndOfficersDataBlockValidator.java]: just tried it and api will fail to parse it if it's not one of the enum values so should be ok. I think it's good practice to have the default so you can leave that in there.

File: src/test/java/uk/gov/companieshouse/overseasentitiesapi/validation/utils/DueDiligenceDataBlockValidatorTest.java (hunk 5 above)
Comment (mattch1) [reply src/test/java/uk/gov/companieshouse/overseasentitiesapi/validation/utils/DueDiligenceDataBlockValidatorTest.java:37]: Added the verification tests

File: src/test/java/uk/gov/companieshouse/overseasentitiesapi/validation/utils/OwnersAndOfficersDataBlockValidatorTest.java (hunk 7 above)
Comment (mattch1) [reply src/test/java/uk/gov/companieshouse/overseasentitiesapi/validation/utils/OwnersAndOfficersDataBlockValidatorTest.java]: Added the verification tests

File: src/test/java/uk/gov/companieshouse/overseasentitiesapi/validation/utils/OwnersAndOfficersDataBlockValidatorTest.java (hunk 6 above)
Comment (mattch1) [reply src/test/java/uk/gov/companieshouse/overseasentitiesapi/validation/utils/OwnersAndOfficersDataBlockValidatorTest.java]: Marked final

Comment (markpit) [reply src/test/java/uk/gov/companieshouse/overseasentitiesapi/validation/utils/OwnersAndOfficersDataBlockValidatorTest.java]: these aren't showing as final for me, maybe missed due to the package move

File: src/test/java/uk/gov/companieshouse/overseasentitiesapi/validation/OwnersAndOfficersDataBlockValidatorTest.java [hunk 8]
```diff
@@ -0,0 +1,230 @@
...
+        List<ManagingOfficerIndividualDto>  managingOfficerIndividualDtoList = new ArrayList<>();
+        managingOfficerIndividualDtoList.add(ManagingOfficerMock.getManagingOfficerIndividualDto());
+        overseasEntitySubmissionDto.setManagingOfficersIndividual(managingOfficerIndividualDtoList);
+        Errors errors = new Errors();
+        ownersAndOfficersDataBlockValidator.validateOwnersAndOfficers(overseasEntitySubmissionDto, errors, LOGGING_CONTEXT);
+        assertError(BENEFICIAL_OWNERS_STATEMENT, String.format("%s for statement that none can be identified", INCORRECTLY_ADDED_BENEFICIAL_OWNER), errors);
+    }
+
+    @Test
+    void testBeneficialOwnerIndividualIsCalled() {
```
Comment (markpit) [inline src/test/java/uk/gov/companieshouse/overseasentitiesapi/validation/OwnersAndOfficersDataBlockValidatorTest.java:165]: on the new 'testXIsCalled' tests can you add the word validator in eg testBeneficialOwnerIndividualValidatorIsCalled

File: src/test/java/uk/gov/companieshouse/overseasentitiesapi/validation/utils/OwnersAndOfficersDataBlockValidatorTest.java (hunk 6 above)
Comment (mattch1) [reply src/test/java/uk/gov/companieshouse/overseasentitiesapi/validation/utils/OwnersAndOfficersDataBlockValidatorTest.java]: Added final

**Review by markpit — APPROVED** [review]
